File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n,m = len(maps),len(maps[0])
    visited = [[-1]*m for _ in range(n)]
    visited[0][0] = 1
    d = [[1,0],[-1,0],[0,-1],[0,1]]

    start_x,start_y = 0,0
    queue = deque([[start_x,start_y]])
    while queue:
        p_x, p_y = queue.popleft()
        for i in range(4):
            next_x = p_x+d[i][0]
            next_y = p_y+d[i][1]

            if (-1<next_x<m) and (-1<next_y<n):
                if visited[next_y][next_x] == -1:
                    if maps[next_y][next_x] == 0:
                        visited[next_y][next_x] = 0
                    elif maps[next_y][next_x] == 1:
                        visited[next_y][next_x] = visited[p_y][p_x] + 1
                        queue.append([next_x,next_y])
                elif visited[next_y][next_x] > 0:
                    if visited[next_y][next_x] > visited[p_y][p_x] + 1:
                        visited[next_y][next_x] = visited[p_y][p_x] + 1
                        queue.append([next_x,next_y])
    return visited[-1][-1]

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        int[][] visited = new int[n][m];
        visited[0][0] = 1;

        int[] controlY = {0, 0, 1, -1};
        int[] controlX = {1, -1, 0, 0};

        Queue<int[]> needVisited = new LinkedList<>();
        needVisited.add(new int[]{0, 0});

        while (!needVisited.isEmpty()) {
            int[] now = needVisited.poll();

            for (int i = 0; i < 4; i++) {
                int nextY = now[0] + controlY[i];
                int nextX = now[1] + controlX[i];
                if (-1 < nextY && nextY < n && -1 < nextX && nextX < m) {
                    if (maps[nextY][nextX] == 1 && visited[nextY][nextX] == 0) {
                        visited[nextY][nextX] = visited[now[0]][now[1]] + 1;
                        needVisited.add(new int[]{nextY, nextX});
                    }
                }
            }
        }

        return visited[n - 1][m - 1] != 0 ? visited[n - 1][m - 1] : -1;
    }
}